include/ndarray/pybind11.h

```diff
@@ -57,188 +175,223 @@ NAMESPACE_BEGIN(detail)
-        _src.reset(src.ptr()); // equivalent to add_ref=true, keep alive for stage 2
-        if (!ndarray::PyConverter< ndarray::Array<T,N,C> >::fromPythonStage1(_src)) {
-            PyErr_Clear();
-            return false;
-        }
-        return true;
-    }
-    static handle cast(const ndarray::Array<T,N,C> &src, return_value_policy /* policy */, handle /* parent */) {
-        return ndarray::PyConverter< ndarray::Array<T,N,C> >::toPython(src);
+        return _helper.init(src) && _helper.check();
     }
+
     void set_value() {
-        if (!ndarray::PyConverter< ndarray::Array<T,N,C> >::fromPythonStage2(_src, _value)) {
-            throw error_already_set();
-        }
+        _value = _helper.convert();
     }
-/* This part is normally created by the PYBIND11_TYPE_CASTER macro, which
- * can't be used here due to the partial specialization
- */
-protected:
-    ndarray::PyPtr _src;
-    ndarray::Array<T,N,C> _value;
-    bool _none = false;
-public:
+
+    static handle cast(const ndarray::Array<T,N,C> &src, return_value_policy /* policy */, handle /* parent */) {
+        return Helper::toPython(src);
+    }
+
     static PYBIND11_DESCR name() { return type_descr(_<ndarray::Array<T,N,C>>()); }
+
     static handle cast(const ndarray::Array<T,N,C> *src, return_value_policy policy, handle parent) {
         return cast(*src, policy, parent);
     }
-    operator ndarray::Array<T,N,C> * () { 
-        if (_none) {
+
+    operator ndarray::Array<T,N,C> * () {
+        if (_helper.isNone) {
             return nullptr;
         } else {
             set_value();
             return &_value;
         }
     }
+
     operator ndarray::Array<T,N,C> & () { set_value(); return _value; }
+
     template <typename _T> using cast_op_type = pybind11::detail::cast_op_type<_T>;
+
+private:
+    ndarray::Array<T,N,C> _value;
+    Helper _helper;
 };
 
 /* @brief A pybind11 type_caster for ndarray::EigenView
  */
 template <typename T, int N, int C, typename Kind, int Rows, int Cols>
 class type_caster< ndarray::EigenView<T,N,C,Kind,Rows,Cols> > {
+    using Helper = ndarray::Pybind11Helper<T,N,C>;
 public:
+
     bool load(handle src, bool) {
-        _none = src.is_none();
-        if (_none) return true;
-        _src.reset(src.ptr()); // equivalent to add_ref=true, keep alive for stage 2
-        if (!ndarray::PyConverter< ndarray::EigenView<T,N,C,Kind,Rows,Cols> >::fromPythonStage1(_src)) {
-            PyErr_Clear();
+        if (!_helper.init(src)) {
             return false;
         }
-        return true;
-    }
-    static handle cast(const ndarray::EigenView<T,N,C,Kind,Rows,Cols> &src, return_value_policy /* policy */, handle /* parent */) {
-        return ndarray::PyConverter< ndarray::EigenView<T,N,C,Kind,Rows,Cols> >::toPython(src);
+        if (_helper.isNone) {
+            return true;
+        }
+        if ((Rows == 1 || Cols == 1) && N == 2) {
+            ndarray::pybind11_np_size_t shape[2] = { -1, -1 };
```

Comment:
So what if it is `size_t`?

Reply:
Good catch; this one needs to be signed even if that's not what pybind11 uses by default.